hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mob/mapreduce/MobRefReporter.java

```diff
@@ -0,0 +1,485 @@
+ *         to impacted rows will still work (e.g. if the HFileLink in our table's mob area was
+ *         manually removed, but the original file in the referenced table is still around for some
+ *         reason.) However, even in this edge case HBase's internal cleaning systems should
+ *         eventually remove that file once it is no longer needed by another table or snapshot.
+ *   * HLINK BUT POINT TO MISSING FILE - There is a pointer in our mob area for this table and CF
+ *         to a file elsewhere on the FileSystem, however the file it points to no longer exists.
+ *   * MISSING FILE BUT IOE WHILE CHECKING HLINKS - We could not find the referenced file, however
+ *         you should check the job logs to see why we couldn't check to see if there is a pointer
+ *         to the referenced file in our archive or another table's archive or mob area.
+ *
+ */
+@InterfaceAudience.Private
+public class MobRefReporter extends Configured implements Tool {
+  private static Logger LOG = LoggerFactory.getLogger(MobRefReporter.class);
+  public static final String NAME = "mobrefs";
+  static final String REPORT_JOB_ID = "mob.report.job.id";
+  static final String REPORT_START_DATETIME = "mob.report.job.start";
+  static final int SCAN_CACHING = 10000;
+  static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+  public static class MobRefMapper extends TableMapper<Text, ImmutableBytesWritable> {
+    @Override
+    public void map(ImmutableBytesWritable r, Result columns, Context context) throws IOException,
+        InterruptedException {
+      if (columns == null) {
+        return;
+      }
+      Cell[] cells = columns.rawCells();
+      if (cells == null || cells.length == 0) {
+        return;
+      }
+      Set<String> files = new HashSet<>();
+      long count = 0;
+      long size = 0;
+      for (Cell c : cells) {
+        if (MobUtils.hasValidMobRefCellValue(c)) {
+          // TODO confirm there aren't tags
+          String fileName = MobUtils.getMobFileName(c);
+          if (!files.contains(fileName)) {
+            context.write(new Text(fileName), r);
+            files.add(fileName);
+          }
+          final int cellsize = MobUtils.getMobValueLength(c);
+          context.getCounter("SIZE_PER_CELL_DIGITS", Integer.toString(Integer.toString(cellsize).length())).increment(1L);
+          size += cellsize;
+          count++;
+        }
+      }
+      context.getCounter("CELLS_PER_ROW_DIGITS", Integer.toString(Long.toString(count).length())).increment(1L);
+      context.getCounter("SIZE_PER_ROW_DIGITS", Integer.toString(Long.toString(size).length())).increment(1L);
+      context.getCounter("MOB","NUM_CELLS").increment(count);
+    }
+  }
+
+  public static class MobRefReducer extends
+      Reducer<Text, ImmutableBytesWritable, Text, Text> {
+
+    TableName table;
+    String mobRegion;
+    Path mob;
+    Path archive;
+    String seperator;
+
+    /* Results that mean things are fine */
```

Comment:
originally I wanted to include this in the result descriptions for the javadocs, but it didn't flow as well there. what do you think, prioritize getting it more visible there?

Reply:
I would say that JavaDoc is quite large as is. I like a really descriptive JavaDoc but wonder if this may be a candidate to have its own documentation somewhere? Otherwise, simply having the fields clearly named as fine/incorrect/missing_data would allow an operator to discover them in the JavaDoc fields section without more prose?

Reply:
I could redo this as a bit in the reference guide on mob. There's no "troubleshooting" section for it yet.

Reply:
Oh also, just updating on this, I think your taxonomy on lines [225](https://github.com/apache/hbase/pull/785/files#diff-cd64a362d79cdf08690a294d3f90feedR225), [229](https://github.com/apache/hbase/pull/785/files#diff-cd64a362d79cdf08690a294d3f90feedR229) and [233](https://github.com/apache/hbase/pull/785/files#diff-cd64a362d79cdf08690a294d3f90feedR233) is nice - only it is code comments and not visible in the Java Docs. Would you think prefixing each Text object with the severity so that the JavaDoc shows its impactl? For example, `HLINK_CLONE`->`OK_HLINK_CLONE`, `ARCHIVE_ERROR_IO`->`INCORRECT_ARCHIVE_ERROR_IO`, `MISSING`->`DATALOSS_MISSING`, etc.?